optex/base/optex.lua

```diff
@@ -483,40 +485,69 @@ end
 -- We use the `node.direct` way of working with nodes. This is less safe, and
 -- certainly not idiomatic Lua, but faster and codewise more close to the way
 -- \TeX/ works with nodes.
---
+local function is_color_needed(head, n, id, subtype) -- returns non-stroke, stroke color needed
+    if id == glyph_id then
+        return true, true
+    end
+    if id == whatsit_id and
+            (subtype == pdfliteral_id
+            or subtype == pdfsave_id
+            or subtype == pdfrestore_id) then
+        return true, true
+    end
+    if id == glue_id then
+        n = getleader(n)
+        if not n then
+            -- ordinary "space" glue
+            return false, false
+        else
+            id = getid(n)
+            if id == hlist_id or id == vlist_id then
+                -- leaders with hlist/vlist get single color
+                return true, false
+            end
+            -- else we have a leader rule checked below (note that color will
+            -- rightly be set before leader, not before the nested rule!)
+        end
+    end
+    if id == rule_id then -- normal or nested under leaders
+        local width, height, depth = getwhd(n)
+        if width <= one_bp or height + depth <= one_bp then
+            return true, true
+        end
+        return true, false
+    end
+    return false, false
+end
+
 local function colorize(head, current, current_stroke)
     for n, id, subtype in traverse(head) do
         if id == hlist_id or id == vlist_id then
             -- nested list, just recurse
             local list = getlist(n)
             list, current, current_stroke = colorize(list, current, current_stroke)
             setlist(n, list)
-        elseif id == glyph_id or id == disc_id or id == rule_id
-                or (id == glue_id and getleader(n))
-                or (id == whatsit_id
-                    and (subtype == pdfliteral_id
-                        or subtype == pdfsave_id
-                        or subtype == pdfrestore_id)) then
-            -- color participating node, check if current color has to change
+        else
+            local nonstroke_needed, stroke_needed = is_color_needed(head, n, id, subtype)
             local new = getattribute(n, color_attribute) or 0
-            if new ~= current then
-                local color = token_getmacro("_color:"..new)
-                head = insertbefore(head, n, pdfliteral(color))
+            local newcolor = nil
+            if current ~= new and nonstroke_needed then
+                newcolor = token_getmacro("_color:"..new)
+                if not newcolor then
+                    print("newcolor", new, newcolor)
```

Comment:
Is it debug-print? It can It can happen if macro level does not cooperates with lua level? But we know that newcolor is nil, we need not print it.

Reply:
At the time of the check, `newcolor` should always be a string. This was a debug print, that could have just as well been `assert(newcolor)`. It is no longer needed. I will delete it soon.